Improve ability to run under non-root user; revoke unneeded capabilities; run under 'sssd' starting F41 by default.

alexey-tikhonov opened the pull request:
tl,dr brief overview of changes:
 - make all folders sssd:sssd owned and group accessible
 - split "sssd.conf -> config.ldb" conversion into two steps: read + write
 - get&apply sssd.conf 'user' and 'debug_level' options right after reading sssd.conf and before doing anything else (so that config.ldb is written under configured service user)
 - in case SSSD is built '--with-sssd-user=sssd' but run under 'root': add 'sssd' group to the process so it can access all sssd:sssd owned folders
 - start with zero (and locked) capabilities whenever possible; otherwise drop all capabilities as soon as possible

For more lengthy and more "user facing" text see commit message of: 
[sssd.service: run under SSSD_USER by default](https://github.com/SSSD/sssd/commit/a481251e07e14670a091f774b3d93faadcc102c9)
  --  this will be a basis for a release note.

There are  a number of known issues remains to address:

 - <s>[SSSDDomain::write()](https://github.com/SSSD/sssd/blob/master/src/config/SSSDConfig/__init__.py.in#L1031) should chown sssd.conf to sssd.service::User, so that ipa-client-install would create sssd.conf readable by SSSD</s> (now included in this PR)
 - https://github.com/SSSD/sssd/issues/5536
 - 'passkey' tests don't work when SSSD runs under 'sssd' (currently just skipped)
 - 'sssd_be' running as non-root fails to execute 'adcli' to renew keytab
 - make 'sssd_kcm' to run under non root by default
 - https://issues.redhat.com/browse/RHEL-11181
 - use 'sysusers' starting f-41
 - `sss_krb5_touch_config()` fails being called by 'sssd_be' via `sss_write_domain_mappings()` or `sss_write_krb5_conf_snippet()`

Above stuff should be addressed before sssd-2.10 release, but this work has been brewing for too long already.
I think it's in a good enough shape for a review.

(plus see additional comments below)

alexey-tikhonov commented:
Now about testing all those different combinations.

CI of this PR itself covers (I) (everything besides f41 and centos-9) and (II.1) (f-41 and centos-9)

#7195 covers (II.2.1) - almost all tests on f-41 and centos-9 and (II.2.2) - core subset of system tests on f-41 and centos-9 (see c4e80942fe071781ecda1b2ff24ab75af57af1a8). Other platforms can be ignored in this PR as it doesn't bring anything new.
#7195 is only a draft with tmp tweaks to get a test run for those combinations. I don't know if we want to keep testing those combinations systematically. On the one hand, I'd like to declare "sssd.conf::user" option deprecated (already in sssd-2.10) and remove it fast (ideally already in f-41 and for sure in c10s). On the other hand, this is what c9s will keep using (if we ever rebase it on sssd-2.10+). And most probably we should keep testing (II.2.1), but this will require a re-work of c4e80942fe071781ecda1b2ff24ab75af57af1a8 and related patches to manage sssd.service::User instead of sssd.conf::user

alexey-tikhonov commented:
A note that there is currently some discrepancy between centos-9 configuration in this PR and a real/projected centos-9 config. This PR configures centos-9 with `User=sssd` while "real" C9S will keep using "User=root" (even once/if rebased on sssd-2.10) This is intentional to have a better CI coverage, at least until there is C10S compose available.

alexey-tikhonov commented:
> CI of this PR itself covers (I) (everything besides f41 and centos-9) and (II.1) (f-41 and centos-9)
> 
> #7195 covers (II.2.1) - almost all tests on f-41 and centos-9 and (II.2.2) - core subset of system tests on f-41 and centos-9

#7195 helped to discover a couple of missing bits:
 - group's access to pubconf/keytab folders
 - read access for 'other' to pidfile (otherwise SSSD can't read leftover pidfile after service user change, might be not a significant problem in the wild, but breaks CI)

Now all CI tests are green in both PRs.

alexey-tikhonov commented:
Rebased and added a patch to `SSSDDomain::write()` to 'chown()' sssd.conf to service user.

alexey-tikhonov commented:
Rebased and added a patch to use 'sysusers' starting Fedora 41

justin-stephenson commented:
Some PRCI jobs use configure sssd with [this action](https://github.com/SSSD/sssd/blob/master/.github/actions/configure/action.yml), does `with-sssd-user=` need to be added to `fedconfig` [here](https://github.com/SSSD/sssd/blob/master/contrib/fedora/bashrc_sssd)?

alexey-tikhonov commented:
> Some PRCI jobs use configure sssd with [this action](https://github.com/SSSD/sssd/blob/master/.github/actions/configure/action.yml), does `with-sssd-user=` need to be added to `fedconfig` [here](https://github.com/SSSD/sssd/blob/master/contrib/fedora/bashrc_sssd)?

Added (and rebased).

justin-stephenson reviewed (approved):
Ack, thank you.

alexey-tikhonov commented on src/conf_macros.m4:
```diff
@@ -35,12 +35,12 @@ AC_DEFUN([WITH_PLUGIN_PATH],
-    config_pidpath="\"VARDIR\"/run"
-    pidpath="${localstatedir}/run"
+    config_pidpath="\"SSS_STATEDIR\"/"
+    pidpath="${localstatedir}/lib/sss/"
```
My understanding is that it only makes sense to put anything (including PID file) into '/run' if any external process is expected to read it there.
So my intention was to stress 'Type=notify' service type (i.e. systemd doesn't watch SSSD PID file, it is for internal use only)
Now having read https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch03s15.html I'm not so sure...
But statement "System programs that maintain transient UNIX-domain sockets must place them in this directory or an appropriate subdirectory" makes me wonder if entire /var/lib/sss/pipes and maybe /mc/ should go to /run as well?

justin-stephenson replied:
I agree about keeping the pid file in /run/, that is what I expect as standard placement for services. I am indifferent about /var/lib/sss/pipes and /mc/ moving

alexey-tikhonov replied:
Thinking more about it... Is PID file required at all?
IIRC, the only use is to check at startup if SSSD is already running.
This can be done iterating over '/proc', similar to https://github.com/SSSD/sssd/blob/master/src/util/find_uid.c#L251

alexey-tikhonov replied:
Anyway, I moved PID file location to '/run/sssd/' in the latest update.

sumit-bose commented on src/monitor/monitor.c:
```diff
@@ -1973,46 +1918,86 @@ static void monitor_restart_service(struct mt_svc *svc)
+    }
+    monitor = talloc_zero(tmp_ctx, struct mt_ctx);
+    if (monitor == NULL) {
+        return 2;
```
Hi,

we currently do not call `talloc_free(tmp_ctx)` when exiting after an error. Do you think it is worth to add it for the sake of a clean process termination?

bye,
Sumit

alexey-tikhonov replied:
Done.

sumit-bose replied:
Thanks.

sumit-bose commented on src/monitor/monitor.c:
Hi,

I think it would be worth to add the exit code overview to the `sssd` man page as well.

bye,
Sumit

alexey-tikhonov replied:
I think it's better to print/provide error description right when it happens (`ERROR`/`sss_log()`)
(I added a couple of missing messages in the latest update).
I think it's more user friendly (specific error description instead of vague "configuration error") and easier to maintain.
But if you are sure it's worth let me know and I'll try add something.

alexey-tikhonov replied:
Added a brief overview.
But I really don't think it makes sense to describe every code like "can't get current capabilities - libcap call failed".
It will be difficult to maintain...

sumit-bose replied:
Hi,

thank you, yes, that's what I had in mind.

bye,
Sumit

sumit-bose commented on src/util/util.h:
```diff
@@ -109,7 +109,6 @@ extern int socket_activated;
 #define FLAGS_INTERACTIVE 0x0002
 #define FLAGS_PID_FILE 0x0004
-#define FLAGS_GEN_CONF 0x0008
 #define FLAGS_NO_WATCHDOG 0x0010
```
Maybe a comment saying that flag `0x0008` aka `1<<3` was removed and can be reused in future would be a good reminder. Or if you think that it should not be reused in future, a comment would be good as well.

alexey-tikhonov replied:
Done.

sumit-bose replied:
Thanks

pbrezina commented on src/confdb/confdb_setup.c:
```diff
@@ -44,6 +44,35 @@
 
 
+errno_t confdb_read_ini(TALLOC_CTX *mem_ctx,
+                     const char *config_file,
+                     const char *config_dir,
+                     bool allow_missing_file,
+                     struct sss_ini **_ini)
```
Indentation.

alexey-tikhonov replied:
Done.

pbrezina commented on src/confdb/confdb_setup.c:
```diff
@@ -44,6 +44,35 @@
+        return ENOMEM;
+    }
+
+    ret = sss_ini_read_sssd_conf(*_ini,
+                                 config_file,
+                                 config_dir);
```
This can maybe go to single line and still fit characters limit?

alexey-tikhonov replied:
Done.

pbrezina commented on src/util/sss_ini.c:
```diff
@@ -395,7 +395,7 @@ int sss_ini_get_int_config_value(struct sss_ini *self,
 
-const char *sss_ini_get_string_config_value(struct sss_ini *self,
+char *sss_ini_get_string_config_value(struct sss_ini *self,
                                             int *error)
```
Indentation.

alexey-tikhonov replied:
Done.

pbrezina commented on src/util/sss_ini.h:
```diff
@@ -136,7 +136,7 @@ int sss_ini_get_int_config_value(struct sss_ini *self,
  */
-const char *sss_ini_get_string_config_value(struct sss_ini *self,
+char *sss_ini_get_string_config_value(struct sss_ini *self,
                                             int *error);
```
Indentation.

alexey-tikhonov replied:
Done.

pbrezina commented on src/util/check_file.c:
```diff
@@ -90,14 +90,16 @@ static errno_t perform_checks(const char *filename,
     if (uid != (uid_t)(-1) && stat_buf->st_uid != uid) {
-        DEBUG(SSSDBG_TRACE_LIBS, "File '%s' must be owned by uid [%d].\n",
-              filename, uid);
+        DEBUG(SSSDBG_TRACE_LIBS,
```
Please change commit subject of `DEBUG: changed verbosity of: ` it does not work well.

alexey-tikhonov replied:
Done.

pbrezina commented on src/confdb/confdb.c:
```diff
@@ -649,8 +649,6 @@ int confdb_init(TALLOC_CTX *mem_ctx,
     struct confdb_ctx *cdb;
```
Please, change commit subject of `MONITOR: changed startup logic to: ` it does not work well.

alexey-tikhonov replied:
Done.

pbrezina commented on src/tests/system/tests/test_passkey.py:
```diff
@@ -82,6 +82,10 @@ def test_passkey__register__ipa(ipa: IPA, moduledatadir: str, testdatadir: str):
 @pytest.mark.importance("critical")
 @pytest.mark.topology(KnownTopologyGroup.AnyProvider)
 @pytest.mark.builtwith(client="passkey", provider="passkey")
+@pytest.mark.require(
```
Can you move this into a function? E.g.

```python
def passkey_require_root(client: Client) -> tuple[bool, str] | bool:
    if client.svc.get_property("sssd", "User") == "root":
        return False, "Passkey tests don't work if SSSD runs under non-root"

    return True

@pytest.mark.require(passkey_require_root)
```

alexey-tikhonov replied:
Done.

alexey-tikhonov replied:
```
@pytest.mark.require.with_args(passkey_requires_root)
```
doesn't work either:
```
tests/test_passkey.py::test_passkey__su (ldap) FAILED 
```

sumit-bose commented on contrib/sssd.spec.in:
```diff
@@ -1050,6 +1047,13 @@ getent passwd sssd >/dev/null || useradd -r -g sssd -d / -s /sbin/nologin -c "Us
+%__rm -f %{mcpath}/sid
+%__chown %{sssd_user}:%{sssd_user} %{dbpath}/*
+%__chown %{sssd_user}:%{sssd_user} %{_sysconfdir}/sssd/sssd.conf
+%__chown -R %{sssd_user}:%{sssd_user} %{_sysconfdir}/sssd/conf.d/*
```
Hi,

I got the error
```
/usr/bin/chown: cannot access '/etc/sssd/conf.d/*': No such file or directory
warning: %post(sssd-common-9.pr7193-04224.fc39.x86_64) scriptlet failed, exit status 1
```
Using `-f` will suppress the error message but the command will still return `1'. Since you are using `-R` and afaik we do not support sub-directories in `conf.d` I wonder if you meant
```
%__chown -R %{sssd_user}:%{sssd_user} %{_sysconfdir}/sssd/conf.d
```
which would chown the `conf.d` directory itself as well.

bye,
Sumit

pbrezina reviewed (changes requested):
Hi, thank you for the patches, it looks good and seems to work well. I have just few minor comments.

alexey-tikhonov commented:
@sumit-bose, @pbrezina, thank you for the review.

I either addresses or replied to all the comments (and also rebased), so this is ready for the next round.

alexey-tikhonov commented on src/tests/system/tests/test_passkey.py:
```diff
@@ -82,6 +89,7 @@
 @pytest.mark.importance("critical")
 @pytest.mark.topology(KnownTopologyGroup.AnyProvider)
 @pytest.mark.builtwith(client="passkey", provider="passkey")
+@pytest.mark.require(lambda client: passkey_test_requires_root(client))
```
It's purpose to capture 'client' into the scope, otherwise name isn't visible.

pbrezina replied:
No, it is not needed. If yes, then it is a bug.

pbrezina commented:
I  have fixed the runtime issue of Debian container and there is an error about creating /run/sssd which seems to be related to this pull request.

alexey-tikhonov commented:
> I have fixed the runtime issue of Debian container and there is an error about creating /run/sssd which seems to be related to this pull request.

hm...
```
/usr/bin/mkdir -p \
...
    /tmp/sssd-intg.ChBL7uWV/sbin \
    /tmp/sssd-intg.ChBL7uWV/share/man \
    /run/sssd \
    /tmp/sssd-intg.ChBL7uWV/lib/sssd \
    /tmp/sssd-intg.ChBL7uWV/lib/ldb \
```
but in the makefile it's all the same:
```
        $(MKDIR_P) \
...
    $(DESTDIR)$(sbindir) \
    $(DESTDIR)$(mandir) \
    $(DESTDIR)$(pidpath) \          <-
    $(DESTDIR)$(pluginpath) \
    $(DESTDIR)$(libdir)/ldb \
```

alexey-tikhonov commented:
> I have fixed the runtime issue of Debian container and there is an error about creating /run/sssd which seems to be related to this pull request.

Fixed.

sumit-bose commented on contrib/sssd.spec.in:
```diff
@@ -1,7 +1,7 @@
 
 # define SSSD user
-%if 0%{?rhel}
+%if 0%{?fedora} >= 41 || 0%{?rhel}
```
Hi,

wouldn't it be more convenient to have this `%if` only once and e.g. set `%global use_sssd_user` which can be checked in the other places? This would it make more easy to understand which other setting might be important for running as SSSD user.

bye,
Sumit

alexey-tikhonov replied:
Good idea. Done.

sumit-bose commented on contrib/sssd.spec.in:
```diff
@@ -1032,7 +1030,7 @@ install -D -p -m 0644 contrib/sssd.sysusers %{buildroot}%{_sysusersdir}/sssd.con
+%if 0%{?fedora} >= 41 || 0%{?rhel}
 %pre common
 %if %{use_sysusers}
 %sysusers_create_compat contrib/sssd.sysusers
```
This is not strictly related to the patch set, but `%sysusers_create_compat contrib/sssd.sysusers` does not work.

The current working directory at this point is the `rpmbuild` directory and both `BUILD` and `BUILDROOT` are already cleared at the the scriptlets are created. So I think the only solution is an additional `SOURCE` file as described in https://docs.fedoraproject.org/en-US/packaging-guidelines/UsersAndGroups/#_dynamic_allocation.

@pbrezina, this means basically to revert 736430aa0e .

bye,
Sumit

alexey-tikhonov replied:
Right, it didn't built, that's why I postponed this (listed in 'TODO').

pbrezina replied:
Hmm, I would bet that it did work for me. Anyway,source file is not needed and it would be the same problem anyway. We need to install the file to sysusersdir (which we do) and then use that one for %sysusers_create_compat.

sumit-bose replied:
Since it is on the TODO list I'll close this thread as 'resolved'.

sumit-bose commented:
Hi,

maybe something for the TODO list:
```
Mar 12 11:32:37 client.test sssd[1182]: (2024-03-12 11:32:37): [be[ipa.test]] [sss_krb5_touch_config] (0x0020): Unable to change mtime of "/etc/krb5.conf" [13]: Permission denied
Mar 12 11:32:37 client.test sssd[1182]: (2024-03-12 11:32:37): [be[ipa.test]] [sss_write_krb5_conf_snippet] (0x0020): Unable to change last modification time of krb5.conf. Created mappings may not be loaded.
```

`/etc/krb5.conf` is touched to indicate to long running processes to reload the Kerberos configuration to at least refresh `[domain_realms]` and most probably any other iteratable sections.

Most probably this "touch" can be moved to `krb5_child` and/or `ldap_child`.

bye,
Sumit

alexey-tikhonov commented:
> maybe something for the TODO list:

Thanks, added.

> 
> ```
> Mar 12 11:32:37 client.test sssd[1182]: (2024-03-12 11:32:37): [be[ipa.test]] [sss_krb5_touch_config] (0x0020): Unable to change mtime of "/etc/krb5.conf" [13]: Permission denied
> Mar 12 11:32:37 client.test sssd[1182]: (2024-03-12 11:32:37): [be[ipa.test]] [sss_write_krb5_conf_snippet] (0x0020): Unable to change last modification time of krb5.conf. Created mappings may not be loaded.
> ```

sumit-bose commented on src/util/capabilities.c:
```diff
@@ -0,0 +1,208 @@
+static inline const char *cap_flag_to_str(cap_flag_value_t flag)
+{
+    if (flag == CAP_SET) {
+        return "*1*";
```
Just out of curiosity, is there a reason for wrapping the `1` in `*`?

alexey-tikhonov replied:
Just to make it more visible during debugging, before I added an option to print only-non-zero caps.

I can remove entire helper (`cap_flag_to_str()`)

sumit-bose replied:
> Just to make it more visible during debugging, before I added an option to print only-non-zero caps.
> 
> I can remove entire helper (`cap_flag_to_str()`)

No, please keep it.

sumit-bose commented:
Hi,

I think I'm basically done with the review. I tested various user setting and they worked as expected. Typically I had to change the ownership of the cache and log files if SSSD was running under a different user (or remove the files before start). I didn't test socket activated services.

Besides my comments about the log level of the capabilities info in `krb5_child` and `ldap_child` and the chown of the log files in the spec file I'm fine.

bye,
Sumit

alexey-tikhonov commented on contrib/sssd.spec.in:
```diff
@@ -1050,6 +1050,13 @@ getent passwd sssd >/dev/null || useradd -r -g sssd -d / -s /sbin/nologin -c "Us
+%__rm -f %{mcpath}/sid
+%__chown %{sssd_user}:%{sssd_user} %{dbpath}/*
+%__chown %{sssd_user}:%{sssd_user} %{_sysconfdir}/sssd/sssd.conf
+%__chown -R %{sssd_user}:%{sssd_user} %{_sysconfdir}/sssd/conf.d
```
I can add this.

And/or add a chown to sssd.service...

Currently there is:
```
$ grep chown *
sssd-autofs.service.in:ExecStartPre=+-/bin/chown @SSSD_USER@:@SSSD_USER@ @logpath@/sssd_autofs.log
sssd-ifp.service.in:ExecStartPre=+-/bin/chown @SSSD_USER@:@SSSD_USER@ @logpath@/sssd_ifp.log
sssd-pac.service.in:ExecStartPre=+-/bin/chown @SSSD_USER@:@SSSD_USER@ @logpath@/sssd_pac.log
sssd-pam.service.in:ExecStartPre=+-/bin/chown @SSSD_USER@:@SSSD_USER@ @logpath@/sssd_pam.log @logpath@/p11_child.log
sssd-ssh.service.in:ExecStartPre=+-/bin/chown @SSSD_USER@:@SSSD_USER@ @logpath@/sssd_ssh.log @logpath@/p11_child.log
sssd-sudo.service.in:ExecStartPre=+-/bin/chown @SSSD_USER@:@SSSD_USER@ @logpath@/sssd_sudo.log
```
  -- all socket activated services, besides 'sssd-nss' (and this is incosistent) 'chown' log files...

Or 'chown' in spec-file and remove all 'chown' from *.serivce files...

alexey-tikhonov replied:
Added 'chown' to spec-file and 'sssd-nss.service'.

alexey-tikhonov commented:
> Typically I had to change the ownership of the cache

At upgrade it should be taken care of by
```
%__chown %{sssd_user}:%{sssd_user} %{dbpath}/*
```

In general, '%{dbpath}' is group-writeable, so SSSD process could delete cache file in case of ownership mismatch even if service user was changed (doesn't matter root->sssd or sssd->root). As it currently happens with config.ldb
But I'm not sure it's worth the trouble...

sumit-bose commented:
> > Typically I had to change the ownership of the cache
> 
> At upgrade it should be taken care of by
> 
> ```
> %__chown %{sssd_user}:%{sssd_user} %{dbpath}/*
> ```
> 

Hi,

yes, that's working fine but I was changing the running user in `sssd.conf` or systemd service file of the installed version of SSSD.

> In general, '%{dbpath}' is group-writeable, so SSSD process could delete cache file in case of ownership mismatch even if service user was changed (doesn't matter root->sssd or sssd->root). As it currently happens with config.ldb But I'm not sure it's worth the trouble...

No, I think this is not need, additionally the cache files should not be removed automatically without the user knowing it will happen, renaming might be ok but even this might cause trouble for an offline system.

bye,
Sumit

alexey-tikhonov commented on src/providers/krb5/krb5_child.c:
```diff
@@ -4159,7 +4160,19 @@ int main(int argc, const char *argv[])
+        talloc_free(caps);
+    } else {
+        DEBUG(SSSDBG_IMPORTANT_INFO, "Failed to get current capabilities\n");
+    }
```
Fixed.

alexey-tikhonov commented:
(rebase)

sumit-bose reviewed (approved):
Hi,

thank you very much for your persistence and patience on this task, ACK.

bye,
Sumit

alexey-tikhonov commented:
Pushed PR: https://github.com/SSSD/sssd/pull/7193

* `master`
    * cb4dbea61980d7e7ca49d1a99e034944e0e3c0d9 - SPEC/SYSTEMD: try harder making sure logs ownership matches service user
    * e37a8c7893f7dee0402ab2439b644fb7b87a795c - MAN: 'monitor' exit codes description
    * 128777896c26ba9c37d888209b65db4648b512ae - MONITOR: free 'tmp_ctx' in case of failure too
    * d45b85b7cf1f4f1401bc5d41682c2082861b6287 - SSSDConfig: chown() sssd.conf to SSSD service user
    * 869ee96526771133b1b18c4f228329d42f5e2e67 - SPEC: build Fedora >= 41 package with sssd user support
    * 07f00135f56a20e3642dfa904965d8c82df7b5dc - TESTS: system: skip 'passkey' tests if SSSD runs under non-root
    * 9eed3873a397713759aa0a1455323fe49c58a0b5 - SPEC: make conf folder g+rx
    * 84c3034dc58e74dbd6c557df1373a3c0ae07a226 - SYSTEMD: set "SecureBits=noroot noroot-locked"
    * e2c26e810c3635124255e7619272591eab143553 - SPEC: replace SUID bit with more fine-grained capabilities
    * 29b1e474ca42ede4ac3f0cbc55de9c1754a2e39e - UTILS: make pidfile readable by everyone
    * 6ca4e47223fcb68b0680294f1463a8e044a2a96b - CONF: store pid file in /run/sssd
    * 583ea7f2dfac15b56ef36742faad8149fda92cf2 - SYSTEMD: remove "PIDFile="
    * 9fbaf6d74bb916ab2db68da409dcd652f8acf1aa - SSS_INI: only check file ownership from 'sssd'
    * b88d56a39c957c71b3280b4b42b07882e35d6123 - SYSTEMD: KCM capabilities
    * aa7cddfa9452a12c9cf79adcad282c4cc30630ea - SPEC: make sure config files are accesible
    * 4c42ca7a9d69ad1bd09ded68de8ba0319d4a9ff7 - SPEC: make sure cache files are accessible
    * 2a59991bed2a77c2e4ee9cb6938e817c60629823 - sssd.service: run under SSSD_USER by default
    * 0e2ed444e6284c2bf6704dd3213044e36e9162bb - KRB5_/LDAP_CHILD: print capabilities at startup
    * 1ea6965c9a1f045513ee39f8d9ef7b274551b800 - MONITOR: startup logic was changed
    * 304fe754119e333516c89839584839485b0ee5d4 - SYSTEMD: responders do not need any capabilities
    * 5bd52025ead1cb7321b4116012cc7f3661866e4d - SYSTEMD: remove unused CAP_KILL
    * 0d686b5d71848dce7757cadc4e67ad6b95b34355 - TOOLS: remove the upgrade-cache command
    * d7042fed2c0bbe74ceac8d4dab50ed75511d9c9e - DEBUG: a couple of message changes
    * a05b025067b1168e7e4ba5f6e32bffb7cdf15619 - SSS_INI: remove 'const' specifier from getter
    * fd23a94ffda6851138a74b4eaba150e353ee638b - MONITOR: move nscd check code to a function
    * 34f7c2eac7def14ab7d700276c141366a0db15c6 - MONITOR: move keyring setup code to a function
    * e306d93f9acc1769a889e7124967fbcb7bbd2a44 - MONITOR: remove unused mt_ctx::conf_path
    * 87b77a011d4d9cc95e70a328f18ee42329ce616c - MONITOR: no need to read domain list twice
    * b1cbf5f59b33a819c9f64e11901ab633fc00a3fa - CONFDB: always delete old ldb-file
    * cff8e1f9914c3edb94f7e9eb7a89d8eba3a2fc06 - CONFDB: split confdb_setup() into 2 steps
    * 8d1b3ef7e3be6b90533ffd7b8d1167a490681364 - SSS_INI: const correctness
    * 4a44cca401dbec16593b93b5c4333005cf6165f4 - Get rid of `--genconf` and `--genconf-section` monitor options.
    * f4ad8c2abab79a4c9224e1813cbb35d982f608d7 - UTILS: add capabilities management helpers
    * 60853c6fabecea628c8e411a310c24b8b2c1eb67 - NSS: don't `fchown()` mem-cache files
    * 52fa441b90c157d5b397a3e5e7a15aa73374d118 - Make all SSSD processes a member of sssd supplementary group.
    * 521f88ef87fa693f07ba48d99fdfb161891b9af6 - SPEC: make '%{pipepath}/private' sssd:sssd owned
    * 47da0b6bcdadc40a13998d8c2a4660f771bd9cb7 - SPEC: make most folders group accessible
    * 102c30a57b27f5bad408412022821d5faa73771c - MONITOR: get rid of unsed FLAGS_GEN_CONF definition
    * d79e0e74e45ae61cb94a26707af2a964ac06f263 - MNITOR: cosmetics
    * 419120f4a514ac7159a575c130b5feaa5a381d36 - MONITOR: replace fprintf() with ERROR()
    * 40cea81b1f3ec98c2f1a71d5c0b5c69dd0797c80 - MONITOR: remove 'opt_netlinkoff' removal notice
    * 40e5309a06e7adea81f9327a0989ee59ca9c4bcd - MONITOR: remove useless trailing '\'

alexey-tikhonov commented:
Hi @sumit-bose,

> maybe something for the TODO list:
> 
> ```
> Mar 12 11:32:37 client.test sssd[1182]: (2024-03-12 11:32:37): [be[ipa.test]] [sss_krb5_touch_config] (0x0020): Unable to change mtime of "/etc/krb5.conf" [13]: Permission denied
> Mar 12 11:32:37 client.test sssd[1182]: (2024-03-12 11:32:37): [be[ipa.test]] [sss_write_krb5_conf_snippet] (0x0020): Unable to change last modification time of krb5.conf. Created mappings may not be loaded.
> ```
> 
> `/etc/krb5.conf` is touched to indicate to long running processes to reload the Kerberos configuration to at least refresh `[domain_realms]` and most probably any other iteratable sections.

This "touch" is done in two functions:
(1) `sss_write_krb5_conf_snippet()`
(2) `sss_write_domain_mappings()`

What is the reason for (1)? It looks like `sss_write_krb5_conf_snippet()` write static, domain independent configs.
 
> Most probably this "touch" can be moved to `krb5_child` and/or `ldap_child`.

I tries to follow code logic to figure out when functions (1) and (2) are executed.
It seems both are triggered by a long chain initiated by DPM_DOMAINS_HANDLER:: ipa/ad_subdomains_handler_send() and ipa/ad_subdomains_init()
It sounds plausible that  `ldap_child` should be used in this chain (to get TGT) but so far I wasn't able to identify where this happens :-/

sumit-bose commented:
> Hi @sumit-bose,
> 
> > maybe something for the TODO list:
> > ```
> > Mar 12 11:32:37 client.test sssd[1182]: (2024-03-12 11:32:37): [be[ipa.test]] [sss_krb5_touch_config] (0x0020): Unable to change mtime of "/etc/krb5.conf" [13]: Permission denied
> > Mar 12 11:32:37 client.test sssd[1182]: (2024-03-12 11:32:37): [be[ipa.test]] [sss_write_krb5_conf_snippet] (0x0020): Unable to change last modification time of krb5.conf. Created mappings may not be loaded.
> > ```
> >     
> >   
> > `/etc/krb5.conf` is touched to indicate to long running processes to reload the Kerberos configuration to at least refresh `[domain_realms]` and most probably any other iteratable sections.
> 
> This "touch" is done in two functions: (1) `sss_write_krb5_conf_snippet()` (2) `sss_write_domain_mappings()`
> 
> What is the reason for (1)? It looks like `sss_write_krb5_conf_snippet()` writes static, domain independent configs.

Hi,

yes, currently it is not expected that the settings written by `sss_write_krb5_conf_snippet()` will change during the runtime of SSSD. But at least for the first time the touch would be needed if not `sss_write_domain_mappings()` would be called immediately afterwards. Maybe both functions can be called from a single function which does the touch at the end.

> 
> > Most probably this "touch" can be moved to `krb5_child` and/or `ldap_child`.
> 
> I tried to follow code logic to figure out when functions (1) and (2) are executed. It seems both are triggered by a long chain initiated by DPM_DOMAINS_HANDLER:: ipa/ad_subdomains_handler_send() or ipa/ad_subdomains_init() It sounds plausible that `ldap_child` should be used in this chain (to get TGT) but so far I wasn't able to identify where this happens :-/

Yes, ti happens during startup and when updating the information about subdomains. I would not couple this with getting the TGT because during  ipa/ad_subdomains_handler_send() you already have a TGT.

bye,
Sumit

alexey-tikhonov commented:
> > This "touch" is done in two functions: (1) `sss_write_krb5_conf_snippet()` (2) `sss_write_domain_mappings()`
> > What is the reason for (1)? It looks like `sss_write_krb5_conf_snippet()` writes static, domain independent configs.
> 
> yes, currently it is not expected that the settings written by `sss_write_krb5_conf_snippet()` will change during the runtime of SSSD. But at least for the first time the touch would be needed if not `sss_write_domain_mappings()` would be called immediately afterwards. Maybe both functions can be called from a single function which does the touch at the end.

Practically that's already (almost) the case: it's either `ad_subdom_reinit()` or `ipa_subdom_reinit()` that calls both (1) and (2)


> > > Most probably this "touch" can be moved to `krb5_child` and/or `ldap_child`.
> > I tried to follow code logic to figure out when functions (1) and (2) are executed. It seems both are triggered by a long chain initiated by DPM_DOMAINS_HANDLER:: ipa/ad_subdomains_handler_send() or ipa/ad_subdomains_init() It sounds plausible that `ldap_child` should be used in this chain (to get TGT) but so far I wasn't able to identify where this happens :-/
> Yes, ti happens during startup and when updating the information about subdomains. I would not couple this with getting the TGT because during ipa/ad_subdomains_handler_send() you already have a TGT.

@sumit-bose, what calls `krb5_child` or `ldap_child` in this chain then?

sumit-bose commented:
> 
> > > I tried to follow code logic to figure out when functions (1) and (2) are executed. It seems both are triggered by a long chain initiated by DPM_DOMAINS_HANDLER:: ipa/ad_subdomains_handler_send() or ipa/ad_subdomains_init() It sounds plausible that `ldap_child` should be used in this chain (to get TGT) but so far I wasn't able to identify where this happens :-/
> 
> > Yes, ti happens during startup and when updating the information about subdomains. I would not couple this with getting the TGT because during ipa/ad_subdomains_handler_send() you already have a TGT.
> 
> @sumit-bose, what calls `krb5_child` or `ldap_child` in this chain then?

Hi,

nothing, this has to be added if one of those helpers should do the touch.

bye,
Sumit
